api/wasm/cpp/proxy_wasm_impl.h

```diff
@@ -294,6 +273,68 @@ class Context {
   void grpcStreamHandler(StringView service, StringView service_name,
       StringView method_name, std::unique_ptr<GrpcStreamHandlerBase> handler);
 
+private:
+  friend class GrpcCallHandlerBase;
+  friend class GrpcStreamHandlerBase;
+
+  uint32_t id_;
+  std::unordered_map<uint32_t, HttpCallCallback> http_calls_;
+  std::unordered_map<uint32_t, GrpcSimpleCallCallback> simple_grpc_calls_;
+  std::unordered_map<uint32_t, std::unique_ptr<GrpcCallHandlerBase>> grpc_calls_;
+  std::unordered_map<uint32_t, std::unique_ptr<GrpcStreamHandlerBase>> grpc_streams_;
+};
+
+// A context unique for each root_id for a use-case (e.g. filter) compiled into module.
+class RootContext : public ContextBase {
+public:
+  RootContext(uint32_t id, StringView root_id) : ContextBase(id), root_id_(root_id) {}
+  ~RootContext() {}
+
+  StringView root_id() { return root_id_; }
+
+  RootContext* asRoot() override { return this; }
+  Context* asContext() override { return nullptr; }
+
+  // Called once when the VM loads and once when each hook loads and whenever configuration changes.
+  virtual void onConfigure(std::unique_ptr<WasmData> /* configuration */) {}
```

Comment:
A general question, what is the scope of root context sharing? does that mean two modules will have to share the same config as onConfigure is invoked within root context?

Reply:
onConfigure will be called per-filter which could share a root context.  The configuration should be a proto which has a oneof if it needs to distiguish the case.  We should document this.

Reply:
As discussed, each filter/use-case in a VM should have a unique root_id, so they would not need to share configuration (although they could with oneof).